src/hll.c

```diff
@@ -92,6 +127,233 @@ enum {
+
+	functionList = FuncnameGetCandidates(qualifiedFunctionNameList, argumentCount,
+										 argumentList, findVariadics,
+										 findDefaults, missingOK);
+
+	if (functionList == NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_FUNCTION),
+						errmsg("function \"%s\" does not exist", functionName)));
+	}
+	else if (functionList->next != NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_AMBIGUOUS_FUNCTION),
+						errmsg("more than one function named \"%s\"", functionName)));
+	}
+
+	/* get function oid from function list's head */
+	functionOid = functionList->oid;
+	return functionOid;
+}
+
+/*
+ * LookupAggregatorOidByName is a wrapper around FunctionOid that uses a cache
+ * to avoid multiple lookups of aggregate within a single session by their names.
+ */
+static Oid
+LookupAggregatorOidByName(char *aggregateName, int argumentCount)
+{
+	HllAggregateCacheEntry *matchingCacheEntry = NULL;
+	ListCell *cacheEntryCell = NULL;
+
+ 	/* search the cache */
+	foreach(cacheEntryCell, hllAggregateCache)
+	{
+		HllAggregateCacheEntry *cacheEntry = (HllAggregateCacheEntry *)lfirst(cacheEntryCell);
+ 		if (strcmp(aggregateName, cacheEntry->aggregateName) == 0 &&
+ 			argumentCount == cacheEntry->argumentCount)
+		{
+			matchingCacheEntry = cacheEntry;
+			break;
+		}
+	}
+
+ 	/* if not found in the cache, call FunctionOid and put the result in cache */
+	if (matchingCacheEntry == NULL)
+	{
+		MemoryContext oldContext = NULL;
+		Oid hllId = get_extension_oid(HLL_EXTENSION_NAME, false);
+		Oid hllSchemaOid = get_extension_schema(hllId);
+		const char *hllSchemaName = get_namespace_name(hllSchemaOid);
+		Oid aggregateId = FunctionOid(hllSchemaName, aggregateName, argumentCount);
+
+ 		if (aggregateId == InvalidOid)
+		{
+			/* if operatorId is invalid, return and do not cache its value */
+			return aggregateId;
+		}
+ 		oldContext = MemoryContextSwitchTo(CacheMemoryContext);
+ 		matchingCacheEntry = palloc0(sizeof(HllAggregateCacheEntry));
+		matchingCacheEntry->aggregateId = aggregateId;
+		matchingCacheEntry->aggregateName = aggregateName;
+		matchingCacheEntry->argumentCount = argumentCount;
+
+		hllAggregateCache = lappend(hllAggregateCache, matchingCacheEntry);
+ 		MemoryContextSwitchTo(oldContext);
+	}
+
+ 	return matchingCacheEntry->aggregateId;
+}
+
+/*
+ * MaximizeCostOfHashAggregateWithHLL maximizes the cost of the path if it tries
+ * to run hll aggregate function with hash aggregate.
+ */
+static void
+MaximizeCostOfHashAggregateWithHLL(Path *path)
+{
+	List *varList = pull_var_clause((Node*) path->pathtarget->exprs,
+									PVC_INCLUDE_AGGREGATES);
+	ListCell *varCell = NULL;
+
+	foreach(varCell, varList)
+	{
+		Var *var = (Var *) lfirst(varCell);
+
+		if (nodeTag(var) == T_Aggref)
+		{
+			Aggref *aggref = (Aggref *) var;
+			int argCount = list_length(aggref->args);
+			Oid addFunctionId = LookupAggregatorOidByName(HLL_ADD_AGG_NAME, argCount);
+			Oid unionFunctionId = LookupAggregatorOidByName(HLL_UNION_AGG_NAME, 1);
```

Comment:
This would be very unlikely to change while the extension is created. Consider caching those Oid's in _pg_init.  In any case I would be happier if we take them out of the loop.